tensorflow/compiler/xla/service/platform_util.cc

```diff
@@ -48,9 +51,14 @@ string CanonicalPlatformName(const string& name) {
   if (platform_str == "cpu") {
     platform_str = "host";
   }
-  // "gpu" and "cuda" mean the same thing.
+  // When configured on CUDA, "gpu" and "cuda" mean the same thing.
+  // When configured on ROCm, "gpu" and "rocm" mean the same thing.
   if (platform_str == "gpu") {
+#if GOOGLE_CUDA
     platform_str = "cuda";
```

Comment:
In hindsight, it was probably not a good idea to use "gpu" and "cuda" synonymously in the XLA codebase. As far as I can tell, this was done for convenience. We already renamed gpu_compiler to nvptx_compiler a while ago. Overloading the term "gpu" to mean both, "cuda" and "rocm" will likely cause confusion. Would we run into issues without it?

Reply:
@thomasjoerg Why this change was added is because:

- At https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/platform_util.cc#L141 , `se::Platform::Name()` is checked to return the StreamExecutor Platform.
- On CUDA platform the implementation is registered with the name "CUDA": https://github.com/tensorflow/tensorflow/blob/master/tensorflow/stream_executor/cuda/cuda_platform.cc#L68
- On ROCm platform the implementation is registered with the name "ROCm": https://github.com/tensorflow/tensorflow/blob/master/tensorflow/stream_executor/rocm/rocm_platform.cc#L30

Reply:
Thanks for the pointers. I'm unclear what or who specifies the platform name "gpu" rather than "cuda" or "rocm", which seems less ambiguous. Would it be an option to discourage usage of "gpu" in favor for the more specific names?

Reply:
Let me post additional code traces here. Essentially it boils down to how we define `/job:worker/replica:0/task:0/device:XLA_GPU:0` be mapped to what StreamExecutor `Platform`.

Reply:
One example is:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/tests/sample_file_test.cc#L38

Some other places in XLA test framework:
- https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/tests/build_defs.bzl#L264
- https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/tests/test_macros.cc#L96
- https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/tests/test_macros.h#L113

The notion of "gpu" is deeply welded into XLA test framework so I chose to change it here.

Reply:
Thanks for the pointers! I agree that the notion of "gpu" is deeply integrated into the testing framework. However, the assumption here is that "gpu" really means NVidia GPU. I'm worried that breaking this assumption could cause issues. For instance,
`"""Removes "gpu" from a backend list if CUDA is not enabled.`
does not really make sense for AMDGPU. We should either carefully review the testing framework to identify such cases, or introduce "amdgpu" or "rocm" as a first-class citizen. Am I making sense?

Reply:
I agree we need to take a deeper look at what "gpu" means but I think that's beyond the scope of this PR. There are outstanding PRs for XLA frontend / backend logic for AMDGPU. Once they are reviewed /  merged, additional PRs to alter XLA unit tests would be filed to cope with differences between NVPTX and AMDGPU.

Reply:
You said that the notion of "gpu" is deeply welded into the XLA test framework, and hence, this PR makes it a synonym of "rcom". However, that will not be enough to make tests work, as the example above shows. What is your motivation to land this change? What functionality will start working (reliably) once it's in?

Reply:
@thomasjoerg let me review unit test results once again and get back to you on this one.

Reply:
sorry for replying this so late. As dependent PRs gradually being merged this PR should need to be revisited.

at `PyLocalClient::Get`, `gpu` is passed to `PlatformUtil::GetPlatform`. Without making `gpu` becoming `rocm` after `PlatformUtil::CanonicalPlatformName`, ROCm StreamExecutor won't be retrieved:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/python/local_client.cc#L171

Reply:
+ @cheshire onto this PR.